Zend/zend_API.c

```diff
@@ -2694,6 +2694,10 @@ ZEND_API zend_result zend_register_functions(zend_class_entry *scope, const zend
 		internal_function->scope = scope;
 		internal_function->prototype = NULL;
 		internal_function->attributes = NULL;
+		ZEND_MAP_PTR_NEW(internal_function->run_time_cache);
```

Comment:
But ... zend_declare_class_constant_ex is also doing `ZEND_MAP_PTR_NEW(ce->mutable_data)` - code which is also invoked when a module declaring class constants is loaded via dl() at run-time. Or is that broken as well?

Reply:
I think it's broken as well :D (or I'm wrong about `ZEND_MAP_PTR_NEW`).

Here `ZEND_MAP_PTR_NEW()` is only called for internal classes with AST constants (e.g. internal enums), which are probably rare, which would explain why we haven't seen the problem yet. Also I've just noticed that dl()ing an extension with internal enums crashes for an other reason.

Reply:
Damn :-D I was looking at that code, saw it, and thought … okay, must be safe then.
However, how would one then fix this category of potential problems? I mean … doing a check for MODULE_TEMPORARY each time run_time_cache is to be accessed is obviously not viable for performance reasons.

Reply:
We would only need to do this check before `ZEND_MAP_PTR_NEW()`.

Apparently `ZEND_MAP_PTR_GET` / `ZEND_MAP_PTR_SET` can handle heap-allocated pointers. In this case the map ptr is a real pointer, not a `CG(map_ptr_base)` offset: https://github.com/php/php-src/blob/9e2de4c2d95b0a391f85121d0d077c8d6d5fdcae/Zend/zend_map_ptr.h#L49-L52

Here we could do a `ZEND_MAP_PTR_INIT(internal_function->run_time_cache, zend_arena_alloc(...))` instead of `ZEND_MAP_PTR_NEW()` + `ZEND_MAP_PTR_SET()` when `EG(active)`.

Reply:
An alternative would be to use a `.reserved` slot, or  to introduce a new `zend_internal_function` field.

The run_time_cache / map_ptr mechanism is useful for user functions because they are immutable. We don't really need this for internal functions.

Reply:
Applied the change, I guess looks good now?

Reply:
Note that installed observers are different from request to request, so this needs run-time mutable data, hence the map_ptr mechanism. The big benefit of the map_ptr mechanism is that it's just an offset into some run-time memory, so that one can easily jump from the immutable (internal function) data structure to mutable run-time data. A `.reserved` slot or new fixed field on `zend_internal_function` does not suffice in itself for this purpose.

Further, it perfectly aligns with the existing userland run_time_cache mechanism, reusing the same variables, saving extra conditions, and having also the same access pattern for internal and userland run_time_caches does nothing but make the devs life a bit easier.

Additionally, at least while writing extension code, I also had the need to attach runtime information to internal functions. The workaround is doing a hashtable lookup, but obviously, this has a performance penalty.
The usefulness for core itself is probably a bit more limited, but the run-time overhead of initializing run_time cache is avoided if cache size is basically zero.

Reply:
> A .reserved slot or new fixed field on zend_internal_function does not suffice in itself for this purpose; they would also have to contain offsets into some per request to be allocated structure, essentially ending up being what map_ptrs currently are anyway providing.

I'm not sure about this. map_ptr is needed for structures that may live in SHM, because we don't store pointers to non-SHM in SHM. But zend_internal_functions leave on the heap and don't have that restriction. We can store a pointer to an array of observers in a zend_internal_function, and modify the array independently in each process or request.

I don't object though, with the ZEND_MAP_PTR_SET change this looks good to me :+1: